providers/implementations/signature/rsa.c

```diff
@@ -0,0 +1,962 @@
+            return 0;
+    }
+
+    if (ctx->mgf1_mdname[0] != '\0') {
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, ctx->mgf1_mdname,
+                                  ctx->mgf1_mdprops);
+
+        if (md == NULL)
+            return 0;
+        EVP_MD_free(ctx->mgf1_md);
+        ctx->mgf1_md = md;
+    }
+
+    return 1;
+}
+
+static int setup_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        return 1;
+    if ((ctx->tbuf = OPENSSL_malloc(RSA_size(ctx->rsa))) == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    return 1;
+}
+
+static void clean_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        OPENSSL_cleanse(ctx->tbuf, RSA_size(ctx->rsa));
+}
+
+static void free_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        OPENSSL_clear_free(ctx->tbuf, RSA_size(ctx->rsa));
+}
+
+static int rsa_sign(void *vprsactx, unsigned char *sig, size_t *siglen,
+                    size_t sigsize, const unsigned char *tbs, size_t tbslen)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+    int ret;
+
+    if (sig == NULL) {
+        *siglen = RSA_size(prsactx->rsa);
+        return 1;
+    }
+
+    if (sigsize < (size_t)RSA_size(prsactx->rsa))
+        return 0;
+
+    if (!setup_md(prsactx))
+        return 0;
+
+    if (prsactx->md != NULL) {
+        if (tbslen != prsactx->mdsize) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST_LENGTH);
+            return 0;
+        }
+
+        if (EVP_MD_type(prsactx->md) == NID_mdc2) {
+            unsigned int sltmp;
+
+            if (prsactx->pad_mode != RSA_PKCS1_PADDING) {
+                ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_PADDING_MODE,
+                               "only PKCS#1 padding supported with MDC2");
+                return 0;
+            }
+            ret = RSA_sign_ASN1_OCTET_STRING(0, tbs, tbslen, sig, &sltmp,
+                                             prsactx->rsa);
+
+            if (ret <= 0) {
+                ERR_raise(ERR_LIB_PROV, ERR_LIB_RSA);
+                return 0;
+            }
+            ret = sltmp;
+        } else if (prsactx->pad_mode == RSA_X931_PADDING) {
+            if ((size_t)RSA_size(prsactx->rsa) < tbslen + 1) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_KEY_SIZE_TOO_SMALL);
+                return 0;
+            }
+            if (!setup_tbuf(prsactx)) {
+                ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+                return 0;
+            }
+            memcpy(prsactx->tbuf, tbs, tbslen);
+            prsactx->tbuf[tbslen] = RSA_X931_hash_id(EVP_MD_type(prsactx->md));
+            ret = RSA_private_encrypt(tbslen + 1, prsactx->tbuf,
+                                      sig, prsactx->rsa, RSA_X931_PADDING);
+            clean_tbuf(prsactx);
+        } else if (prsactx->pad_mode == RSA_PKCS1_PADDING) {
+            unsigned int sltmp;
+
+            ret = RSA_sign(EVP_MD_type(prsactx->md), tbs, tbslen,
```

Comment:
RSA_sign() is full of stuff you dont want to do directly..
i.e:- encode_pkcs1()

Reply:
Not sure how else to produce a signature. What output do you expect? The way it's currently designed, the produced signature is a DER encoded X509_SIG. The only way I can see that change is to make signatures a first class provider side object, and use serializers to encode them.

Reply:
(comment removed, it was utter bollocks... I should not comment before morning coffee...)

Reply:
I'm not sure how all that is supposed to be done, er, indirectly. I'm also not quite sure what kind of output is expected, it seems a bit different depending on the algo

Reply:
I suppose that since that will be fixed by #9138, we can stop worrying here.